Лабораторная работа 4

flamangoo opened the pull request.

flamangoo committed:
Лабораторная работа 4

andreiivanov-wa commented on lab_4/task_2.py:
```diff
@@ -0,0 +1,30 @@
+
+def get_percents(char_dictionary):  # функция принимает словарь символов
+    sum_of_values = sum(char_dictionary.values())
+    dictionary_with_percents = {}
```
можно просто назвать переменную percents 
Вы же не называете функцию function_for_getting_percents

flamangoo replied:
@andreiivanov-wa в попытке №2 ответ был отправлен в виде пулл реквеста, ссылка была, но попытка засчитана как неудачная. Нужно отправлять просто ссылку без лишних комментариев?

![image](https://user-images.githubusercontent.com/115110092/198687266-8f7c0e23-7fb0-464f-9a46-c2de8d504c2c.png)

andreiivanov-wa replied:
в ответе не было ссылки на пулреквест, преподаватель же не помнит все ваши прошлые пулреквесты. Вы, возможно, изменили свой пул реквест, но в СДО в ответе ссылку не указали, куда перейти и посмотреть исправления.
![image](https://user-images.githubusercontent.com/27698506/198694698-da7057bd-1821-4619-bed2-34b35854a755.png)

andreiivanov-wa replied:
это мне нужно было открыть ваши предыдущие ответы, найти в них ссылку на старый пулреквест, проверить, что вы обновили его. кто-то новый пулоеквест создает, кто-то изменяет старый, но преподавателю вместо того, чтобы в этом разбираться легче просто не засчитать попытку, оформленную не по правилам )

andreiivanov-wa replied:
а, там можно развернуть комментарий и ниже будет ссылка...
так я эти комментарии и не читал никогда и ожидал там увидеть ссылку, без лишних действий. Пойди догадайся еще, что надо крестик нажать, который вообще сверху текста нарисован

andreiivanov-wa replied:
![image](https://user-images.githubusercontent.com/27698506/198695452-bd85a68c-bae0-4a34-91aa-c07b76451bdc.png)

andreiivanov-wa replied:
вы отправляейте просто ссылку на пулреквест, чтобы она не затерялась в комментариях, а комментарии здесь пишите, если что-то хотите спросить или уточнить

andreiivanov-wa commented on lab_4/task_3.py:
```diff
@@ -0,0 +1,14 @@
+    if index is None:
+        index = len(list_)
+    if index == len(list_):
+        f_delete = list_[:index - 1]
```
1) путаница в логике
a) если вызвать delete(list_, len(list_) - 1) удалится последний элемент и отработает код в строке 7
b) если вызвать delete(list_, len(list_)) то тоже удалится последний элемент, но отработает код в строке 5
c) если вызвать delete(list_, len(list_) + 1) то ничего не удалится и отработает код в строке 7
случаи b) и c) не должны отличаться, если запрашивается удаление индекса который >= длины списка, то ничего не должно удаляться
2) код в строке 5 -- это частный случай кода в строке 7. нет необходимости два раза писать и общий случай, и частный случай. если None достаточно просто задать index = len(list_) - 1 и код в общем виде в строке 7 отработает корректно

andreiivanov-wa commented on lab_4/task_3.py:
```diff
@@ -0,0 +1,14 @@
+
+print(delete([0, 0, 1, 2], index=0))  # [0, 1, 2]
+print(delete([0, 1, 1, 2, 3], index=1))  # [0, 1, 2, 3]
+print(delete([0, 1, 2, 3, 4, 4]))  # [0, 1, 2, 3, 4]
```
print(delete([1, 2, 3], index=-1)) # [1, 2, 1, 2, 3]
нужно корректно обработать случай отрицательных индексов

andreiivanov-wa replied:
здесь как раз можно подумать на тему, как лучше задать параметр по умолчанию, я бы сделал
def delete(list_, index=-1)
и в случае если index < 0 подумать, как корректно переопределить индекс в строке 3, чтобы код return в конце функции отработал в общем виде корректно и удалял последний символ для  -1, предполедний символ для -2 и т.д. Подсказка:
если для index=-1 переопределяет index = len(list_) - 1, то для общего случая
index < 0 переопределим index = ?

andreiivanov-wa commented on lab_4/task_2.py:
```diff
@@ -0,0 +1,30 @@
+def get_count_char(str_):  # функция принимает строку
+    str_ = str_.lower()
+    default_count = 0
```
а почему введена переменная default_count = 0, но не введена переменная default_increment = 1 и тогда в 7 строке
char_dictionary[char] = char_dictionary.get(char, default_count) + default_increment 
лишние переменные тоже не надо вводить, достаточно написать
char_dictionary[char] = char_dictionary.get(char, 0) + 1

flamangoo replied:
> 

Поняла. Я сделала так, как было указано в лекции, инкремент там вводили как переменную, не обратила внимания. Спасибо

andreiivanov-wa commented on lab_4/task_2.py:
```diff
@@ -0,0 +1,30 @@
+    sum_of_values = sum(char_dictionary.values())
+    dictionary_with_percents = {}
+    for char in char_dictionary:
+        dictionary_with_percents[char] = round(char_dictionary[char] / sum_of_values * 100, 2)
```
на подумать (можно не реализовывать): в таком алгоритме сумма процентов для всех символов может запросто получиться меньше 100, а ожидается, что процентное содержание будет равно 100
например дляdict_ = {'a': 33333, 'b': 33333, 'c': 33334} получим проценты {'a': 33.33, 'b': 33.33, 'c': 33.33} что в сумме даст 99.99. Потерялся 0.01 процент. Если реализуете так, чтобы в сумме было 100, поставлю + доп балл для итоговой оценки

flamangoo committed:
Update task_2.py

flamangoo committed:
Update task_3.py

flamangoo committed:
Update task_3.py

flamangoo committed:
Update task_2.py

Попробовала реализовать точность для чисел с плавающей точкой с помощью метода math.fsum. Если числа не округлять, то метод отслеживает "потерянные" цифры во время суммирования и получается 100.0

flamangoo committed:
Update task_2.py

Выполнено дополнительное задание

flamangoo committed:
Update task_2.py

andreiivanov-wa commented on lab_4/task_2.py:
```diff
@@ -0,0 +1,32 @@
+    for char in char_dictionary:
+        percents[char] = char_dictionary[char] / sum_of_values * 100
+    error = 100 - sum(percents.values())  # нахождение отклонения от 100%
+    percents.update({'err': error})  # добавим ошибку как дополнительный ключ
```
здесь лучше создать ключ более уникальный, чтобы не было символов, например 
percents[''] = error # лучше так, чем update
update полезен когда много значений словарем отправляешь, а когда одно -- зачем создавать лишний словарь, чтобы передать его в функцию update

flamangoo replied:
И правда, я об этом не подумала. Спасибо!

flamangoo committed:
Update task_2.py
